src/dataflow/src/server.rs

```diff
@@ -417,26 +418,18 @@ impl PendingPeek {
                             copies += diff;
                         }
                     });
-                    if copies < 0 {
+                    let copies: usize = if copies < 0 {
                         return Err(format!(
                             "Invalid data in source, saw retractions ({}) for row that does not exist: {:?}",
                             copies * -1,
                             &*borrow,
                         ));
-                    }
-
-                    // When we have a LIMIT we can restrict the number of copies we make.
-                    // This protects us when we have many copies of the same records, as
-                    // the DD representation uses a binary count and may not exhaust our
-                    // memory in situtations where this copying might.
-                    if let Some(limit) = max_results {
-                        let limit = std::convert::TryInto::<isize>::try_into(limit);
-                        if let Ok(limit) = limit {
-                            copies = std::cmp::min(copies, limit);
-                        }
-                    }
-                    for _ in 0..copies {
-                        results.push(result.clone());
+                    } else {
+                        copies.try_into().unwrap()
+                    };
+                    // if copies > 0 ... otherwise skip
+                    if let Some(copies) = NonZeroUsize::new(copies) {
+                        results.push((result, copies));
                     }
 
                     // If we hold many more than `max_results` records, we can thin down
```

Comment:
We can change that test, sure.

But I don't want to change the below code (which sorts and then truncates above max_results), because:

(1) That code is just an optimization anyway; it doesn't need to be precise; it's fine if we retain more rows than we need, and since the worst case is when every count is 1 anyway, we won't be doing anything worse than we were before.

(2) Doing it precisely requires a full scan of the counts after sorting, to find where to cut.